pkg/aws/eni: Adding support for prefix delegation in AWS bare metal

41ks opened the pull request:
# What does this do

This PR aims to add the possibility of prefix delegation on AWS instances of type `bare metal`. The original idea was to allow this feature only on instances of Hypervisor type `nitro` (PR #18463). However, this also works on nitro based instances of type bare metal. The only instances that don't support this feature are those of Hypervisor type `xen`.

This feature is also available on AWS's own CNI [ENABLE_PREFIX_DELEGATION](https://github.com/aws/amazon-vpc-cni-k8s?tab=readme-ov-file#enable_prefix_delegation-v190).

# How was it tested

This change was successfully tested on our clusters. Having noticed a problem with AWS nitro based bare metal instances not being assigned IP prefixes, this fix solved our issue.

```release-note
AWS ENI IPAM Mode: Added support for prefix delegation on AWS bare metal instances
```

HadrienPatte commented on pkg/ipam/types/types.go:
```diff
@@ -25,6 +25,9 @@ type Limits struct {
 	// delegation are supported on an instance. Bare metal instances would have empty string.
 	HypervisorType string
+
+	// IsBareMetal tracks wether an instance is a bare metal instance or not
```
```suggestion
	// IsBareMetal tracks whether an instance is a bare metal instance or not
```

liyihuang commented on pkg/aws/eni/node.go:
```diff
@@ -789,8 +789,8 @@ func (n *Node) IsPrefixDelegated() bool {
-	// Allocating prefixes is supported only on nitro instances
-	if limits.HypervisorType != "nitro" {
+	// Allocating prefixes is supported only on nitro and bare metal instances
+	if limits.HypervisorType != "nitro" && !limits.IsBareMetal {
```
thanks for the change. Do you mind to add the test case for it?

41ks replied:
Thank you for the review. I have added test cases to cover this logic.

I don't work for Isovalent, so I don't have access to the notion page you sent.

liyihuang replied:
Sorry, it's from wrong clipboard. I was trying to paste where the unit test for prefix is.

liyihuang reviewed (approved):
Thanks for the contribution. Overall it looks right to me. Hope you can add the unit test for this logic if possible

tklauser reviewed (changes requested):
@41ks please run `make generate-k8s-api && make manifests` to update the generated deepequal method (see https://github.com/cilium/cilium/actions/runs/15391313276/job/43537611720?pr=39678) and please also squash all the commits into a single one. Thanks!

41ks committed:
pkg/aws/eni: Adding support for prefix delegation in AWS bare metal

This adds support for prefix delegation on AWS bare metal instances by checking against the BareMetal flag provided by the AWS API.

Signed-off-by: Alex Melhem <alex.melhem@datadoghq.com>

tklauser reviewed (approved):
Thanks
